pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -389,15 +415,79 @@ func (b *Base) Observables(rsrc interface{}, labels map[string]string) []reconci
 
 // Objects for Messaging service
 func (s *Messaging) Objects(rsrc interface{}, rsrclabels map[string]string, observed, dependent, aggregated []reconciler.Object) ([]reconciler.Object, error) {
-	r := rsrc.(*alpha1.CDAPMaster)
-	return getStatefulServiceResources(&r.Spec.Messaging.CDAPStatefulServiceSpec, rsrc, rsrclabels, alpha1.ServiceMessaging)
+	master := rsrc.(*alpha1.CDAPMaster)
+	var upgradeJob *batchv1.Job
+	for _, item := range reconciler.ObjectsByType(observed, k8s.Type) {
+		if job, ok := item.Obj.(*k8s.Object).Obj.(*batchv1.Job); ok {
+			if master.Status.UpgradeJobStartTimeNano != 0 && strings.Contains(job.Name, strconv.FormatInt(master.Status.UpgradeJobStartTimeNano, 10)) {
+				upgradeJob = job
+				break
+			}
+		}
+	}
+	var upgradeResources []reconciler.Object
+	if upgradeJob == nil && master.Status.UpgradeJobStartTimeNano == 0 &&
+		master.Status.ExistingImage != "" && master.Status.ExistingImage != master.Spec.Image {
+		// launch job
+		master.Status.UpgradeJobStartTimeNano = time.Now().UnixNano()
+		master.Status.IsUpgrading = true
+		var err error
+		upgradeResources, err = getUpgradeResources(
+			&alpha1.UpgradeJobSpec{
+				Image:    master.Spec.Image,
+				JobName:  getUpgradeJobName(master.Status.UpgradeJobStartTimeNano),
+				HostName: getConfigName(master, "router"),
+			}, rsrclabels)
+		if err != nil {
+			master.Status.UpgradeJobStartTimeNano = 0
+			master.Status.IsUpgrading = false
+			return nil, err
+		}
+	} else if upgradeJob != nil && upgradeJob.Status.Succeeded == 0 && time.Now().Add(-20*time.Minute).UnixNano() <= master.Status.UpgradeJobStartTimeNano {
+		// job is still running
+		var err error
+		upgradeResources, err = getUpgradeResources(
+			&alpha1.UpgradeJobSpec{
+				Image:    master.Spec.Image,
+				JobName:  getUpgradeJobName(master.Status.UpgradeJobStartTimeNano),
+				HostName: getConfigName(master, "router"),
+			}, rsrclabels)
+		if err != nil {
+			return nil, err
+		}
+	} else if upgradeJob != nil && upgradeJob.Status.Succeeded > 0 {
+		// job finished
+		master.Status.UpgradeJobStartTimeNano = 0
+		master.Status.IsUpgrading = false
+	} else if upgradeJob != nil && (time.Now().Add(-20*time.Minute).UnixNano() > master.Status.UpgradeJobStartTimeNano || upgradeJob.Status.Failed >= 4) {
```

Comment:
I think the controller shouldn't automatically change the Spec. The spec is the desire state as declared by the user, the controller will try its best to make current state to converge to the desire state. The upgrade Job itself can fail due to various reason (including timeout, but the Job should exit itself with fail state, instead of getting killed externally). The controller should reflect failure in the Status and the user/caller can monitor the status and decide what to do if it fails.